lib/spack/llnl/util/tty/log.py

```diff
@@ -445,62 +571,47 @@ def _writer_daemon(self, stdin):
 
         log_file = self.log_file
 
-        def handle_write(force_echo):
-            # Handle output from the with block process.
-            # If we arrive here it means that in_pipe was
-            # ready for reading : it should never happen that
-            # line is false-ish
-            line = in_pipe.readline()
-            if not line:
-                return (True, force_echo)  # break while loop
-
-            # find control characters and strip them.
-            controls = control.findall(line)
-            line = re.sub(control, '', line)
-
-            # Echo to stdout if requested or forced
-            if echo or force_echo:
-                try:
-                    if termios:
-                        conf = termios.tcgetattr(sys.stdout)
-                        tostop = conf[3] & termios.TOSTOP
-                    else:
-                        tostop = True
-                except Exception:
-                    tostop = True
-                if not (tostop and _is_background_tty()):
-                    sys.stdout.write(line)
-                    sys.stdout.flush()
-
-            # Stripped output to log file.
-            log_file.write(_strip(line))
-            log_file.flush()
-
-            if xon in controls:
-                force_echo = True
-            if xoff in controls:
-                force_echo = False
-            return (False, force_echo)
-
         try:
-            with _keyboard_input(stdin):
+            with keyboard_input(stdin) as kb:
                 while True:
+                    # ensure that input settings are sane.
+                    kb.check_fg()
+
                     # No need to set any timeout for select.select
                     # Wait until a key press or an event on in_pipe.
                     rlist, _, _ = select.select(istreams, [], [])
+
                     # Allow user to toggle echo with 'v' key.
                     # Currently ignores other chars.
                     # only read stdin if we're in the foreground
-                    if stdin in rlist and not _is_background_tty():
+                    if stdin in rlist and not _is_background_tty(stdin):
                         if stdin.read(1) == 'v':
                             echo = not echo
 
                     if in_pipe in rlist:
-                        br, fe = handle_write(force_echo)
-                        force_echo = fe
-                        if br:
+                        # Handle output from the calling process.
+                        line = in_pipe.readline()
+                        if not line:
                             break
 
+                        # find control characters and strip them.
+                        controls = control.findall(line)
```

Comment:
Because it's far simpler this way.  `handle_write` was extremely confusing.

1. It did not just handle a write -- It read the line and returned something based on the result that would trigger a break (or not) in the surrounding loop.  That made the control flow hard to follow.
2. It shadowed `force_echo` and returned a value that would modify `force_echo` in the surrounding scope.
3. It doesn't really stand on its own -- you *have* to understand the loop below to understand what it does.

For those reasons I think it just confuses the control flow here.  Add to that that the writer daemon is persistent and manages long-lived state (like `echo` and `force_echo`) as local variables, and I think there is already enough control complexity with the daemon function and the loop.  It makes more sense to have simpler control flow here than separation of concerns/encapsulation.

If anything it might make sense to make the daemon a class instead of a function (so the state management is clearer) -- but I think that is a larger refactor.

Reply:
This is OK (generally I prefer while loops with substantial logic refactor into a while loop with a function call, but the logic here doesn't quite make the threshold of "too much"), but overall these arguments feel like they are targeting symptoms rather than the cause. I think most of these complaints could be resolved by making a separate function with a clear name and an actual docstring.

> It did not just handle a write -- It read the line and returned something based on the result that would trigger a break (or not) in the surrounding loop

That's if the stream ends. This is a common theme of while loops that read streams. I don't think it's overly complex to put in a function that reads a stream and can return the possibility that the stream is finished.

> It shadowed force_echo and returned a value that would modify force_echo in the surrounding scope.

That's because it was defined inside the function rather than as a separate function.

> It doesn't really stand on its own -- you have to understand the loop below to understand what it does.

docstring

Reply:
I'm still not convinced and am not interested in re-introducing what I think is an overly complex function at this point.

Reply:
yeah ok -- I think it makes sense to make this a separate function.

Reply:
This is done.